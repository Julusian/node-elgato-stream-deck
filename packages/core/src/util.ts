import { InternalFillImageOptions } from './models/base'

export interface FillImageTargetOptions {
	colorMode: 'bgr' | 'rgba'
	xFlip?: boolean
	yFlip?: boolean
	rotate?: boolean
}

export function imageToByteArray(
	imageBuffer: Buffer,
	sourceOptions: InternalFillImageOptions,
	targetOptions: FillImageTargetOptions,
	destPadding: number,
	imageWidth: number,
	imageHeight?: number
): Buffer {
	if (!imageHeight) imageHeight = imageWidth

	const byteBuffer = Buffer.alloc(destPadding + imageWidth * imageHeight * targetOptions.colorMode.length)

	const flipColours = sourceOptions.format.substring(0, 3) !== targetOptions.colorMode.substring(0, 3)

	for (let y = 0; y < imageHeight; y++) {
		const rowOffset = destPadding + imageWidth * targetOptions.colorMode.length * y
		for (let x = 0; x < imageWidth; x++) {
			// Apply x/y flips
			let x2 = targetOptions.xFlip ? imageWidth - x - 1 : x
			let y2 = targetOptions.yFlip ? imageHeight - y - 1 : y

			if (targetOptions.rotate) {
				// Swap x and y
				const tmpX = x2
				x2 = y2
				y2 = tmpX
			}

			const srcOffset = y2 * sourceOptions.stride + sourceOptions.offset + x2 * sourceOptions.format.length

			const red = imageBuffer.readUInt8(srcOffset)
			const green = imageBuffer.readUInt8(srcOffset + 1)
			const blue = imageBuffer.readUInt8(srcOffset + 2)

			const targetOffset = rowOffset + x * targetOptions.colorMode.length
			if (flipColours) {
				byteBuffer.writeUInt8(blue, targetOffset)
				byteBuffer.writeUInt8(green, targetOffset + 1)
				byteBuffer.writeUInt8(red, targetOffset + 2)
			} else {
				byteBuffer.writeUInt8(red, targetOffset)
				byteBuffer.writeUInt8(green, targetOffset + 1)
				byteBuffer.writeUInt8(blue, targetOffset + 2)
			}
			if (targetOptions.colorMode.length === 4) {
				byteBuffer.writeUInt8(255, targetOffset + 3)
			}
		}
	}

	return byteBuffer
}

export const BMP_HEADER_LENGTH = 54
export function writeBMPHeader(buf: Buffer, iconSize: number, iconBytes: number, imagePPM: number): void {
	// Uses header format BITMAPINFOHEADER https://en.wikipedia.org/wiki/BMP_file_format

	// Bitmap file header
	buf.write('BM')
	buf.writeUInt32LE(iconBytes + 54, 2)
	buf.writeInt16LE(0, 6)
	buf.writeInt16LE(0, 8)
	buf.writeUInt32LE(54, 10) // Full header size

	// DIB header (BITMAPINFOHEADER)
	buf.writeUInt32LE(40, 14) // DIB header size
	buf.writeInt32LE(iconSize, 18)
	buf.writeInt32LE(iconSize, 22)
	buf.writeInt16LE(1, 26) // Color planes
	buf.writeInt16LE(24, 28) // Bit depth
	buf.writeInt32LE(0, 30) // Compression
	buf.writeInt32LE(iconBytes, 34) // Image size
	buf.writeInt32LE(imagePPM, 38) // Horizontal resolution ppm
	buf.writeInt32LE(imagePPM, 42) // Vertical resolution ppm
	buf.writeInt32LE(0, 46) // Colour pallette size
	buf.writeInt32LE(0, 50) // 'Important' Colour count
}
