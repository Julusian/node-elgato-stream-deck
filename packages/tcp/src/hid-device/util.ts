export class QueuedCommand {
	public readonly promise: Promise<Buffer>
	public readonly commandType: number

	constructor(commandType: number) {
		this.commandType = commandType
		this.promise = new Promise<Buffer>((resolve, reject) => {
			this.resolve = resolve
			this.reject = reject
		})
	}

	resolve(_res: Buffer): void {
		throw new Error('No promise to resolve')
	}

	// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
	reject(_res: any): void {
		throw new Error('No promise to reject')
	}
}
