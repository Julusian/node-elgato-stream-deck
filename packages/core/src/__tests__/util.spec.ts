import { transformImageBuffer } from '../util.js'

function getSimpleBuffer(dim: number, components: 3 | 4): Buffer {
	const buf = Buffer.alloc(dim * dim * components)
	for (let i = 0; i < buf.length; i++) {
		buf[i] = i
	}
	return buf
}
describe('imageToByteArray', () => {
	test('basic rgb -> rgba', () => {
		const srcBuffer = getSimpleBuffer(2, 3)
		const res = transformImageBuffer(
			srcBuffer,
			{ format: 'rgb', offset: 0, stride: 2 * 3 },
			{ colorMode: 'rgba' },
			5,
			2,
			2,
		)
		expect(res).toMatchSnapshot()
	})
	test('basic rgb -> bgr', () => {
		const srcBuffer = getSimpleBuffer(2, 3)
		const res = transformImageBuffer(
			srcBuffer,
			{ format: 'rgb', offset: 0, stride: 2 * 3 },
			{ colorMode: 'bgr' },
			4,
			2,
			2,
		)
		expect(res).toMatchSnapshot()
	})
	test('basic bgra -> bgr', () => {
		const srcBuffer = getSimpleBuffer(2, 4)
		const res = transformImageBuffer(
			srcBuffer,
			{ format: 'bgra', offset: 0, stride: 2 * 4 },
			{ colorMode: 'bgr' },
			4,
			2,
			2,
		)
		expect(res).toMatchSnapshot()
	})
	test('basic bgra -> rgba', () => {
		const srcBuffer = getSimpleBuffer(2, 4)
		const res = transformImageBuffer(
			srcBuffer,
			{ format: 'bgra', offset: 0, stride: 2 * 4 },
			{ colorMode: 'rgba' },
			4,
			2,
			2,
		)
		expect(res).toMatchSnapshot()
	})

	test('basic vflip', () => {
		const srcBuffer = getSimpleBuffer(3, 3)
		const res = transformImageBuffer(
			srcBuffer,
			{ format: 'bgr', offset: 0, stride: 3 * 3 },
			{ colorMode: 'bgr', yFlip: true },
			4,
			3,
			3,
		)
		expect(res).toMatchSnapshot()
	})

	test('basic xflip', () => {
		const srcBuffer = getSimpleBuffer(3, 3)
		const res = transformImageBuffer(
			srcBuffer,
			{ format: 'bgr', offset: 0, stride: 3 * 3 },
			{ colorMode: 'bgr', xFlip: true },
			4,
			3,
			3,
		)
		expect(res).toMatchSnapshot()
	})
})
