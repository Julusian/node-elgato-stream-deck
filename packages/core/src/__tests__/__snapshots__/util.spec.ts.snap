// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`imageToByteArray basic bgra -> bgr 1`] = `
Uint8Array [
  0,
  0,
  0,
  0,
  0,
  1,
  2,
  4,
  5,
  6,
  8,
  9,
  10,
  12,
  13,
  14,
]
`;

exports[`imageToByteArray basic bgra -> rgba 1`] = `
Uint8Array [
  0,
  0,
  0,
  0,
  2,
  1,
  0,
  255,
  6,
  5,
  4,
  255,
  10,
  9,
  8,
  255,
  14,
  13,
  12,
  255,
]
`;

exports[`imageToByteArray basic rgb -> bgr 1`] = `
Uint8Array [
  0,
  0,
  0,
  0,
  2,
  1,
  0,
  5,
  4,
  3,
  8,
  7,
  6,
  11,
  10,
  9,
]
`;

exports[`imageToByteArray basic rgb -> rgba 1`] = `
Uint8Array [
  0,
  0,
  0,
  0,
  0,
  0,
  1,
  2,
  255,
  3,
  4,
  5,
  255,
  6,
  7,
  8,
  255,
  9,
  10,
  11,
  255,
]
`;

exports[`imageToByteArray basic vflip 1`] = `
Uint8Array [
  0,
  0,
  0,
  0,
  18,
  19,
  20,
  21,
  22,
  23,
  24,
  25,
  26,
  9,
  10,
  11,
  12,
  13,
  14,
  15,
  16,
  17,
  0,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
]
`;

exports[`imageToByteArray basic xflip 1`] = `
Uint8Array [
  0,
  0,
  0,
  0,
  6,
  7,
  8,
  3,
  4,
  5,
  0,
  1,
  2,
  15,
  16,
  17,
  12,
  13,
  14,
  9,
  10,
  11,
  24,
  25,
  26,
  21,
  22,
  23,
  18,
  19,
  20,
]
`;
